src/SarifDeferredSample/Program.cs

```diff
@@ -0,0 +1,97 @@
+﻿using Microsoft.CodeAnalysis.Sarif;
+using Microsoft.CodeAnalysis.Sarif.Readers;
+using Newtonsoft.Json;
+using System;
+using System.Diagnostics;
+using System.IO;
+
+namespace SarifDeferredSample
+{
+    // NOTE: Not being submitted. This is a sample of usage of the Deferred Collections.
+
+    internal class Program
+    {
+        private const float BytesPerMB = (float)(1024 * 1024);
+
+        private static void Main(string[] args)
+        {
+            bool deferred = true;
+
+            string filePath = args[0];
+            SarifLog log = null;
+
+            Console.WriteLine($"Loading {filePath}{(deferred ? " (deferred)" : "")}...");
+
+            Measure(() =>
+            {
+                JsonSerializer serializer = new JsonSerializer();
+                serializer.ContractResolver = (deferred ? SarifDeferredContractResolver.Instance : SarifContractResolver.Instance);
+
+                using (JsonTextReader reader = (deferred ? new JsonPositionedTextReader(filePath) : new JsonTextReader(new StreamReader(filePath))))
+                {
+                    log = serializer.Deserialize<SarifLog>(reader);
```

Comment:
>Deserialize [](start = 37, length = 11)

No matter the outcome of this investigation, we should change the usage pattern recommended in [https://github.com/Microsoft/sarif-sdk/blob/master/docs/how-to.md](https://github.com/Microsoft/sarif-sdk/blob/master/docs/how-to.md). It recommends reading the entire file contents into a string, then deserializing with `JsonConvert.DeserializeObject<T>`. Instead, it should use this pattern of creating a reader, creating a serializer, and calling `JsonSerializer.Deserialize<T>`. The current pattern is simpler, but we shouldn't recommend a pattern that uses that much memory.

Reply:
Yes. Using a JsonTextReader instead of File.ReadAllText halves the peak memory use when reading any Sarif.